contracts/zaps/ZapTrancheHop.sol

```diff
@@ -0,0 +1,116 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+import "../interfaces/IYearnVaultV2.sol";
+import "../libraries/Authorizable.sol";
+import "../interfaces/IERC20.sol";
+import "../interfaces/ITranche.sol";
+
+contract ZapTrancheHop is Authorizable {
+    // Store the accessibility state of the contract
+    bool public isFrozen = false;
+    // Tranche factory address for Tranche contract address derivation
+    address internal immutable _trancheFactory;
+    // Tranche bytecode hash for Tranche contract address derivation.
+    // This is constant as long as Tranche does not implement non-constant constructor arguments.
+    bytes32 internal immutable _trancheBytecodeHash;
+
+    /// @param __trancheFactory Address of the TrancheFactory contract
+    /// @param __trancheBytecodeHash Hash of the Tranche bytecode.
+    constructor(address __trancheFactory, bytes32 __trancheBytecodeHash)
+        Authorizable()
+    {
+        _authorize(msg.sender);
+        _trancheFactory = __trancheFactory;
+        _trancheBytecodeHash = __trancheBytecodeHash;
+    }
+
+    /// @dev Requires that the contract is not frozen
+    modifier notFrozen() {
+        require(!isFrozen, "Contract frozen");
+        _;
+    }
+
+    /// @dev Allows an authorized address to freeze or unfreeze this contract
+    /// @param _newState True for frozen and false for unfrozen
+    function setIsFrozen(bool _newState) external onlyAuthorized() {
+        isFrozen = _newState;
+    }
+
+    /// @notice Redeems Principal and Yield tokens and deposits the underlying assets received into
+    /// a new tranche. The target tranche must use the same underlying asset.
+    /// @param _underlying The underlying ERC20 token contract of the yearn vault.
+    /// @param _positionFrom The wrapped position of the originating tranche.
+    /// @param _expirationFrom The expiration timestamp of the originating tranche.
+    /// @param _positionTo The wrapped position of the target tranche.
+    /// @param _expirationTo The expiration timestamp of the target tranche.
+    /// @param _amountPt Amount of principal tokens to redeem and deposit into the new tranche.
+    /// @param _amountYt Amount of yield tokens to redeem and deposit into the new tranche.
+    /// @param _ptExpected The minimum amount of principal tokens to mint.
+    /// @return returns the minted amounts of principal and yield tokens (PT and YT)
+    function hopToTranche(
+        IERC20 _underlying,
+        address _positionFrom,
+        uint256 _expirationFrom,
+        address _positionTo,
+        uint256 _expirationTo,
+        uint256 _amountPt,
+        uint256 _amountYt,
+        uint256 _ptExpected
```

Comment:
I think we want both `ptExpected` and `ytExpected` since there's edge cases where it's possible to get the pt you expect without getting the yt you expect

Reply:
I'm okay with including it defensively, but the reason I excluded it initially is because the `ytMinted` should always be equal to the amount of underlying deposited. and we check that `ytMinted >= balance`